File: Python/array/maximum_product_subarray.py
import sys
from typing import List

def maxProduct(nums: List[int]) -> int:
    result, imax, imin = -sys.maxsize, 1, 1
    for num in nums:
        re1, re2 = imax, imin
        imax = max(re1 * num, re2 * num, num)
        imin = min(re2 * num, re1 * num, num)
        result = max(result, imax)
    
    return result

if __name__ == "__main__":
    print(maxProduct([-4,-3,-2]))

File: core/tmp/Java/src/com/dbc/algorithm/Array/MaximumProductSubarray.java
package com.dbc.algorithm.Array;

public class MaximumProductSubarray {
    public static int maxProduct(int[] nums) {
        int result = Integer.MIN_VALUE, max = 1, min = 1;
        for (int num : nums) {
            int mx = max, mi = min;
            max = Math.max(num, Math.max(num * mx, num * mi));
            min = Math.min(num, Math.min(num * mx, num * mi));
            result = Math.max(result, max);
        }
        return result;
    }
}
